cluster-autoscaler/proposals/plugable-provider-grpc.md

```diff
@@ -0,0 +1,604 @@
+
+## Proposal
+
+There are couple of examples of pluggable designs using Go SDKs that would guide us on how to extend CA
+to support custom providers as plugins.
+This approach is inspired based on [Hashicorp go-plugin](https://github.com/hashicorp/go-plugin) and [Grafana Go SDK for plugins](https://github.com/grafana/grafana-plugin-sdk-go).
+
+The proposed solution will deploy the external grpc provider along side CA and both communicates via gRPC with TLS/SSL. The external provider would be part of a separate deployment, and so we should deploy it independently, as shown in the ![Figure](./images/external-provider-grpc.jpg).
+
+This approach exposes a common gRPC API server.
+
+
+
+The API calls would be defined by the operations exposed by the `CloudProvider` and `NodeGroup` interfaces in [CA](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/cloud_provider.go#L50):
+
+```go
+type CloudProvider interface {
+	Name() string
+
+	NodeGroups() []NodeGroup
+
+	NodeGroupForNode(*apiv1.Node) (NodeGroup, error)
+
+	Pricing() (PricingModel, errors.AutoscalerError)
+
+	NewNodeGroup(machineType string, labels map[string]string, systemLabels map[string]string,
+		taints []apiv1.Taint, extraResources map[string]resource.Quantity) (NodeGroup, error)
+
+	GetResourceLimiter() (*ResourceLimiter, error)
+
+	GPULabel() string
+
+	GetAvailableGPUTypes() map[string]struct{}
+
+	Cleanup() error
+
+	Refresh() error
+
+  ...
+}
+
+type NodeGroup interface {
+	MaxSize() int
+
+	MinSize() int
+
+	TargetSize() (int, error)
+
+	IncreaseSize(delta int) error
+
+	DeleteNodes([]*apiv1.Node) error
+
+	DecreaseTargetSize(delta int) error
+
+	Id() string
+
+	Nodes() ([]Instance, error)
+
+	TemplateNodeInfo() (*schedulerframework.NodeInfo, error)
+
+	Exist() bool
+
+	Create() (NodeGroup, error)
+
+	Delete() error
+
+	Autoprovisioned() bool
+}
+
+```
+
+In the following I detail the main operations of the CloudProvider GRPC service:
+
+```protobuf
+syntax = "proto3";
+
+package clusterautoscaler.cloudprovider.v1;
+
+import "google/protobuf/descriptor.proto";
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/wrappers.proto";
+
+import "k8s.io/apimachinery/pkg/apis/meta/v1/generated.proto";
+import "k8s.io/apimachinery/pkg/api/resource/generated.proto";
+import "k8s.io/api/core/v1/generated.proto"
+
+option go_package = "clusterautoscaler.cloudprovider";
+
+service CloudProvider {
+  // CloudProvider specific RPC functions
+  rpc Name(NameRequest)
+    returns (NameResponse) {}
+
+  rpc NodeGroups(NodeGroupsRequest)
+    returns (GetNameResponse) {}    
+
+  rpc NodeGroupForNode(NodeGroupForNodeRequest)
+    returns (NodeGroupForNodeResponse) {}
+
+  rpc PricingNodePrice(PricingNodePriceRequest)
+    returns (PricingNodePriceResponse) {}
+
+  rpc PricingPodPrice(PricingPodPriceRequest)
+    returns (PricingPodPriceResponse)
+
+  rpc NewNodeGroup(NewNodeGroupRequest)
```

Comment:
@MaciekPytel  I can see this function used by many providers. Is that correct ?

Reply:
It's not called at all by CA logic (except for some old unittests that call it on mock cloudprovider). So even if some provider were to implement it it would do nothing as the function will never be called.

Almost all existing providers return ErrNotImplemented for this method. The only exception is ovhcloud. However, since it was just added I'm guessing this is because we lack documentation for adding providers and the authors of the PR were unaware of the above. cc: @Kafei59 - can you confirm?

Reply:
Alright! I'll remove it.

Reply:
Yep, I only tried to follow the interface contract, but in fact, it if's not called ... well, it's useless :)